timer: kernel: cortex_m_systick:  fix multiple minor issues

avolkov-1221 opened the pull request:
This patchset fix issues described in #75693, also it contains some minor refactorings like unified 'cortex-m-systick.yaml'

avolkov-1221 committed:
timer: cortex_m_systick: add new generic file 'cortex-m-systick.yaml'

Relocate common properties from dedicated SOC's related
Cortex-M SysTick DTS files into the new generic
'cortex-m-systick.yaml' one.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
timer: cortex_m_systick: minor refactoring

Replace useless "TICKLESS" define by
IS_ENABLED(CONFIG_TICKLESS_KERNEL) one, and replace error-prone
'cycle_t' pseudotype defs by real typedefs.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

ithinuel commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -260,85 +309,95 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
+			last_load = MIN_DELAY;
 		} else {
-			last_load = delay;
+			const uint32_t max_cycles = max_ticks * CYC_PER_TICK;
```
IMHO it makes it clearer that this is a constant to have this as an all CAPS `#define` rather than a variable looking snake-case `const`.

avolkov-1221 replied:
I've not objections. Accepting.

ithinuel commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -260,85 +309,95 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
-	uint32_t unannounced = cycle_count - announced_cycles;
-	uint32_t cyc = elapsed() + unannounced;
+	if (IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		k_spinlock_key_t key = k_spin_lock(&lock);
+		uint32_t unannounced = cycle_count - announced_cycles;
+		uint32_t cyc = elapsed() + unannounced;
```
Please be consistent in the style used.

```suggestion
		K_SPINLOCK(&lock) {
			uint32_t unannounced = cycle_count - announced_cycles;
			uint32_t cyc = elapsed() + unannounced;
```

avolkov-1221 replied:
I'd prefer to use K_SPINLOCK(), but it's not always possible due to the nature of the macro implementation. The compiler starts generating false positive warnings about uninitialized variables.

ithinuel commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -421,13 +480,16 @@ void sys_clock_idle_exit(void)
 {
-	SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+	K_SPINLOCK(&lock) {
+		sys_write32(0, SYST_CSR);
```
Unlike the original code, this is also changing the clock source and the interrupt enable.
Any reason for that change?

avolkov-1221 replied:
These bits are re-enabled by the SYST_CSR_RUN macro later, but the missing read op here for TICKINT clearing is a bug.

ithinuel replied:
Alternate clock source isn’t used in Zephyr yet but there is an issue requesting its addition.

avolkov-1221 replied:
> Alternate clock source isn’t used in Zephyr yet but there is an issue requesting its addition.

It was my issue and frozen PR :). This PR is a cutted down version of the previous one.

avolkov-1221 replied:
I mean, is it some delay will be add if CLKSOURCE will be changed in the disabled state of the SysTick ? If not, then using 0 for the inactive CSR would be reasonable, and nasty TIMER_STOPPED can be thrown away. ARM's  public docos tell nothing about this.

BTW I've remember another reason to move away from the current CMSIS based implementation: multicore. In the current implementation, the base address is hardcoded, but it can be wrong in multicore case.

ithinuel replied:
> In the current implementation, the base address is hardcoded, but it can be wrong in multicore case.

It is not wrong. The SYSTick lives on the Private Peripheral Bus that is core specific.
That is to say the SysTick that core0 accesses is a different one that of what core1’s accessing.

avolkov-1221 replied:
Ok, I've fixed the 'last_loaded' related bug. Could you recheck the latest version?

ithinuel commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -421,13 +480,16 @@ void sys_clock_idle_exit(void)
 
 	if (last_load == TIMER_STOPPED) {
-		SysTick->CTRL |= SysTick_CTRL_ENABLE_Msk;
+		sys_write32(SYST_CSR_RUN, SYST_CSR);
```
If I understand correctly, re-enabling systick will cause the counter to resume and on the next systick isr, `elapsed` calculation will be boggus if it falls in any of the case where `last_load` is used.

So, shouldn’t `last_load` also be set to a sound/meaningful value?
Or shouldn’t `elapsed` be fixed to account for the cases where `last_load` represents something else?

avolkov-1221 replied:
Yeah, it's looks like a real major bug. IMHO it's better to stop using the "TIMER_STOPPED" at all. But it will be out of scope this PR.

ithinuel commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -152,25 +186,27 @@ static uint32_t elapsed(void)
-	uint32_t dcycles;
 	uint32_t dticks;
 
+	k_spinlock_key_t key = k_spin_lock(&lock);
```
Inconsistent style.

Also, I am not sure how advisable it is to spinlock in this ISR.
AFAIU, there shoul not be use of these API at higher priority than this ISR.

avolkov-1221 replied:
My fault in comment (the code literally is more then 2 years old), actually it's not for higher level ISR, but mainly for the latest multi-cores Cortexes. IMO it's better to slow down by noop interrupt disable, then debugging race condition after.

ithinuel reviewed (changes requested):
Unfortunately, this has too many changes without a specfic usecase/reproducible bug to address.

The only part that I think could be good are:
- the factorization of the DTS binding
- fixing the isr signature (removing the spurious `arg`)
- adding the `interrupt` attribute ~~although I’m still not convinced this is 100% required~~

avolkov-1221 commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -182,52 +218,65 @@ void sys_clock_isr(void *arg)
-
-		dcycles = cycle_count - announced_cycles;
-		dticks = dcycles / CYC_PER_TICK;
+		dticks = (uint32_t)(cycle_count - announced_cycles);
```
It was my issue and  frozen PR :)

avolkov-1221 committed:
timer: cortex_m_systick: use direct interrupt handler instead of C-function

The current driver implements the global defined "systick" interrupt
callback "sys_clock_isr" as a standard C function with an argument.
However, ARM's direct interrupt handlers do not have any arguments;
they must be declared as "void handler(void)".

Additionally, the direct handler should include some missing special
header/footer.

Fixes: #75693

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

ithinuel reviewed (changes requested):
My previous review still holds.

Unfortunately, this has too many changes without a specfic usecase/reproducible bug to address.

The only part that I think could be good are:

- the factorization of the DTS binding
- fixing the isr signature (removing the spurious arg)
- adding the interrupt attribute

The rest is a matter of style & personal preference.

If something is wrong in the CMSIS API, please report that upstream.
In the mean time, the current implementation is proven to work.

avolkov-1221 commented:
> Unfortunately, this has too many changes without a specfic usecase/reproducible bug to address.
> 
> The only part that I think could be good are:
> 
> * the factorization of the DTS binding
> * fixing the isr signature (removing the spurious `arg`)
> * adding the `interrupt` attribute ~although I’m still not convinced this is 100% required~

Ok, I've reverted the last path from the patchset. Probably it will be better to create  RFC ticket for CMSIS related usage before.

avolkov-1221 commented:
Hi @ithinuel,

Could you pinging the other(s) reviewers to merge/review the patchset ?

Regards
Andrey

avolkov-1221 commented:
Hi @teburd @andyross @galak @anangl,
Ping... to review/merge this PR
